### DISCUSSION TITLE
Removing query matrix seems to offer similar performance

### DISCUSSION THREAD

**Discussion by brandon-xyzw** [KellerJordan/modded-nanogpt#93] [Show and tell]
```
class CausalSelfAttention(nn.Module):
    def __init__(self, dim: int, num_heads: int, max_seq_len: int, head_dim=128):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = head_dim
        hdim = num_heads * head_dim
        std = 0.5 * (dim ** -0.5)
        bound = (3 ** 0.5) * std
        self.kv_w = nn.Parameter(torch.empty(2, hdim, dim).uniform_(-bound, bound))  # Still KV
        self.lambdas = nn.Parameter(torch.tensor([0.5, 0.5]))
        self.rotary = Rotary(head_dim, max_seq_len)
        self.c_proj = CastedLinear(hdim, dim)
        self.c_proj.weight.detach().zero_()

    def forward(self, x: Tensor, ve: Tensor | None, block_mask: BlockMask):
        B, T = x.size(0), x.size(1)
        assert B == 1
        k, v = F.linear(x, self.kv_w.flatten(end_dim=1).type_as(x)).view(B, T, 2 * self.num_heads, self.head_dim).chunk(2, dim=-2)
        q = x.view(B, T, self.num_heads, self.head_dim)  # Q is x, reshaped

        q = norm(q)
        q = self.rotary(q)

        k = norm(k)
        k = self.rotary(k)

        if ve is not None:
            v = self.lambdas[0] * v + self.lambdas[1] * ve.view_as(v)
        else:
            v = self.lambdas[0] * v
        y = flex_attention(q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), block_mask=block_mask, scale=0.12).transpose(1, 2)
        y = y.contiguous().view(B, T, self.num_heads * self.head_dim)
        y = self.c_proj(y)
        return y
```

# Results
## No Q matrix:

step: 1770/1770 val_loss:3.3060 train_time: 167406ms step_avg: 94.58ms
peak memory allocated: 29848 MiB

## QKV:
step:1770/1770 val_loss:3.2828 train_time:172851ms step_avg:97.66ms
peak memory allocated: 30724 MiB

I'm not sure how to test given there seems to be an unresolved versioning issue which is affecting me. Not sure why the repo stops at 1770 instead of the loss we want, but it seems comparable so figured I post about it.

**Comment by linux-leo** [comment]
I think if you remove the q matrix then q norm can also be removed, as x is already normed when passed to the attention layer and q is x.

**Comment by MeetThePatel** [comment]
> Not sure why the repo stops at 1770 instead of the loss we want

The num_iterations is defined here, so you can tune this for your run. https://github.com/KellerJordan/modded-nanogpt/blob/master/train_gpt.py#L450

As for why the runs are for a fixed num_iterations, a couple schedules are based on the assumption of a fixed length training run. For example, the learning rate schedule is defined in terms of the run length: https://github.com/KellerJordan/modded-nanogpt/blob/master/train_gpt.py#L526C1-L533C39 . The window schedule is also based on the run length: https://github.com/KellerJordan/modded-nanogpt/blob/master/train_gpt.py#L539C1-L540C58 .
